File: solutions/python/05/kmp/main.py
def build_next(p: str) -> list:
    m = len(p)
    nxt = [0] * m
    j = 0
    for i in range(1, m):
        while j > 0 and p[i] != p[j]:
            j = nxt[j - 1]
        if p[i] == p[j]:
            j += 1
            nxt[i] = j
    return nxt


def kmp_search(text: str, pat: str) -> list:
    nxt = build_next(pat)
    j = 0
    pos = []
    for i in range(len(text)):
        while j > 0 and text[i] != pat[j]:
            j = nxt[j - 1]
        if text[i] == pat[j]:
            j += 1
        if j == len(pat):
            pos.append(i - len(pat) + 1)
            j = nxt[j - 1]
    return pos


def main() -> None:
    print(kmp_search("ababaabab", "aba"))


if __name__ == "__main__":
    main()

File: solutions/go/05/kmp/main.go
package main

import "fmt"

func buildNext(p string) []int {
	m := len(p)
	nxt := make([]int, m)
	j := 0
	for i := 1; i < m; i++ {
		for j > 0 && p[i] != p[j] {
			j = nxt[j-1]
		}
		if p[i] == p[j] {
			j++
			nxt[i] = j
		}
	}
	return nxt
}

func kmpSearch(text, pat string) []int {
	nxt := buildNext(pat)
	j := 0
	var pos []int
	for i := 0; i < len(text); i++ {
		for j > 0 && text[i] != pat[j] {
			j = nxt[j-1]
		}
		if text[i] == pat[j] {
			j++
		}
		if j == len(pat) {
			pos = append(pos, i-len(pat)+1)
			j = nxt[j-1]
		}
	}
	return pos
}

func main() {
	fmt.Println(kmpSearch("ababaabab", "aba"))
}
